src/databricks/labs/ucx/assessment/dashboards.py

```diff
@@ -0,0 +1,370 @@
+            try:
+                dashboards.append(next(dashboards_iterator))
+            except StopIteration:
+                break
+            except DatabricksError as e:
+                logger.warning("Cannot list next Redash dashboards page", exc_info=e)
+                break
+        return dashboards
+
+    def _get_dashboards(self, *dashboard_ids: str) -> list[SdkRedashDashboard]:
+        dashboards = []
+        for dashboard_id in dashboard_ids:
+            dashboard = self._get_dashboard(dashboard_id)
+            if dashboard:
+                dashboards.append(dashboard)
+        return dashboards
+
+    def _get_dashboard(self, dashboard_id: str) -> SdkRedashDashboard | None:
+        try:
+            return self._ws.dashboards.get(dashboard_id)
+        except DatabricksError as e:
+            logger.warning(f"Cannot get Redash dashboard: {dashboard_id}", exc_info=e)
+            return None
+
+    def _try_fetch(self) -> Iterable[Dashboard]:
+        for row in self._fetch(f"SELECT * FROM {escape_sql_identifier(self.full_name)}"):
+            yield Dashboard(*row)
+
+    def list_queries(self, dashboard: Dashboard | None = None) -> Iterator[Query]:
+        """List queries.
+
+        Args:
+            dashboard (DashboardType | None) : List queries for dashboard. If None, list all queries.
+                Defaults to None.
+
+        Note:
+            This public method does not adhere to the common crawler layout, still, it is implemented to avoid/postpone
+            another crawler for the queries by retrieving the queries every time they are requested.
+        """
+        for query in self._list_legacy_queries(dashboard):
+            yield Query.from_legacy_query(query)
+
+    def _list_legacy_queries(self, dashboard: Dashboard | None = None) -> Iterator[LegacyQuery]:
+        """List legacy queries.
+
+        Args:
+            dashboard (DashboardType | None) : List queries for dashboard. If None, list all queries.
+                Defaults to None.
+
+        Note:
+            This public method does not adhere to the common crawler layout, still, it is implemented to avoid/postpone
+            another crawler for the queries by retrieving the queries every time they are requested.
+        """
+        if dashboard:
+            queries_iterator = self._list_legacy_queries_from_dashboard(dashboard)
+        else:
+            queries_iterator = self._list_all_legacy_queries()
+        # Redash APIs are very slow to paginate, especially for large number of dashboards, so we limit the listing
+        # to a small number of items in debug mode for the assessment workflow just to complete.
+        counter = itertools.count()
+        while self._debug_listing_upper_limit is None or self._debug_listing_upper_limit > next(counter):
+            try:
+                yield next(queries_iterator)
+            except StopIteration:
+                break
```

Comment:
This can also be simplified using `itertools.islice()`.

Reply:
https://github.com/databrickslabs/ucx/pull/3397#discussion_r1877602313